controllers/wso2is_controller.go

```diff
@@ -0,0 +1,592 @@
+	} else if err != nil {
+		log.Error(err, "Failed to get Deployment")
+		return ctrl.Result{}, err
+	}
+
+	// Ensure the deployment size is the same as the spec
+	size := instance.Spec.Size
+	foundReplicas := found.Spec.Replicas
+
+	instance.Status.Replicas = fmt.Sprint(*foundReplicas)
+
+	if *found.Spec.Replicas != size {
+		found.Spec.Replicas = &size
+		err = r.Update(ctx, found)
+		if err != nil {
+			log.Error(err, "Failed to update Deployment", "Deployment.Namespace", found.Namespace, "Deployment.Name", found.Name)
+			return ctrl.Result{}, err
+		}
+		// Spec updated - return and requeue
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Update the IS status with the pod names
+	// List the pods for this IS's deployment
+	podList := &corev1.PodList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(instance.Namespace),
+		client.MatchingLabels(labelsForWso2IS(instance.Name)),
+	}
+	if err = r.List(ctx, podList, listOpts...); err != nil {
+		log.Error(err, "Failed to list pods", "WSO2IS.Namespace", instance.Namespace, "WSO2IS.Name", instance.Name)
+		return ctrl.Result{}, err
+	}
+	podNames := getPodNames(podList.Items)
+
+	// Update status.Nodes if needed
+	if !reflect.DeepEqual(podNames, instance.Status.Nodes) {
+		instance.Status.Nodes = podNames
+		err := r.Status().Update(ctx, &instance)
+		if err != nil {
+			log.Error(err, "Failed to update WSO2IS status")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// labelsForWso2IS returns the labels for selecting the resources
+// belonging to the given WSO2IS CR name.
+func labelsForWso2IS(depname string) map[string]string {
+	return map[string]string{
+		"deployment": depname,
+		"app":        depname,
+		"monitoring": "jmx",
+		"pod":        depname,
+	}
+}
+
+// getPodNames returns the pod names of the array of pods passed in
+func getPodNames(pods []corev1.Pod) []string {
+	var podNames []string
+	for _, pod := range pods {
+		podNames = append(podNames, pod.Name)
+	}
+	return podNames
+}
+
+// addNamespace adds a new NameSpace
+func (r *Wso2IsReconciler) addNamespace(m wso2v1beta1.Wso2Is) corev1.Namespace {
+	namespace := corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   m.Namespace,
+			Labels: labelsForWso2IS(m.Name),
+		},
+	}
+	ctrl.SetControllerReference(&m, &namespace, r.Scheme)
+	return namespace
+}
+
+// addServiceAccount adds a new ServiceAccount
+func (r *Wso2IsReconciler) addServiceAccount(m wso2v1beta1.Wso2Is) *corev1.ServiceAccount {
+	svc := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "wso2svc-account",
+			Namespace: m.Namespace,
+		},
+	}
+	ctrl.SetControllerReference(&m, svc, r.Scheme)
+	return svc
+}
+
+// add adds a new PersistentVolume
+func (r *Wso2IsReconciler) addPersistentVolumeClaim(m wso2v1beta1.Wso2Is) *corev1.PersistentVolumeClaim {
+	pvc := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "user-store-pv-claim",
```

Comment:
Shall we move all these hardcoded names to constants? then we can refer the resource names using those constants

Reply:
Done